snowflake/snowflake_test.go

```diff
@@ -0,0 +1,230 @@
+	assert.Error(err)
+	assert.Contains(err.Error(), "The maximum life cycle of the snowflake algorithm is 179 years")
+}
+
+func Test_SetResolver_Nil(t *testing.T) {
+	assert := assert.New(t)
+	g := &Generator{}
+	err := g.setResolver(nil)
+	assert.Error(err)
+}
+
+func Test_SetResolver_Ok(t *testing.T) {
+	assert := assert.New(t)
+	g := &Generator{}
+	err := g.setResolver(AtomicResolver)
+	assert.NoError(err)
+	err = g.setResolver(AtomicResolver)
+	assert.NoError(err)
+}
+
+func Test_SetNodeIDProvider_Nil(t *testing.T) {
+	assert := assert.New(t)
+	g := &Generator{
+		Settings: Settings{
+			Epoch:            time.Now(),
+			SequenceResolver: AtomicResolver,
+		},
+	}
+	err := g.setNodeIDProvider(nil)
+	assert.Error(err)
+}
+
+func Test_SetNodeIDProvider_Ok(t *testing.T) {
+	g := &Generator{
+		Settings: Settings{
+			Epoch:            time.Now(),
+			SequenceResolver: AtomicResolver,
+		},
+	}
+
+	assert := assert.New(t)
+	err := g.setNodeIDProvider(customProvider)
+	assert.NoError(err)
+
+	generated, err := g.NextID()
+	assert.NoError(err)
+	decomposed := g.Decompose(generated)
+	assert.Equal(nodeID, decomposed.NodeID)
+
+	err = g.setNodeIDProvider(customProviderPlus)
+	assert.NoError(err)
+	generated, err = g.NextID()
+	assert.NoError(err)
+	decomposed = g.Decompose(generated)
+	assert.Equal(nodeID, decomposed.NodeID)
+}
+
+func newID(t *testing.T) uint64 {
+	assert := assert.New(t)
+
+	id, err := NextID()
+	if err != nil {
+		t.Fatal("ID not generated")
+	}
+	assert.NotZero(id)
+	return id
+}
+
+func currentTime() time.Time {
+	return common.CurrentUTCTime()
+}
+
+func Test_NextIDFor10Sec(t *testing.T) {
+	assert := assert.New(t)
+	sequenceBits := 8
+	var numID uint32
+	var lastID uint64
+	var maxSequence uint16
+	nodeID, err := host.Lower16BitPrivateIP()
+	assert.NoError(err)
+
+	initial := currentTime()
+	current := initial
+	for current.Sub(initial) < 10*time.Second {
+		id := newID(t)
+		parts := Decompose(id)
+		numID++
+
+		if id <= lastID {
+			t.Fatal("duplicated ID")
+		}
+		lastID = id
+
+		current = currentTime()
+
+		actualMSB := parts.MSB
+		if actualMSB != 0 {
+			t.Errorf("unexpected msb: %d", actualMSB)
+		}
+
+		actualTime := parts.Timestamp
+		overtime := actualTime.Sub(current)
+		if overtime > 0 {
+			t.Errorf("unexpected overtime: %d", overtime)
+		}
+
+		actualSequence := parts.Sequence
+		if maxSequence < actualSequence {
+			maxSequence = actualSequence
+		}
+
+		actualNodeID := parts.NodeID
+		if actualNodeID != nodeID {
+			t.Errorf("unexpected nodeID: %d", actualNodeID)
+		}
+	}
+
+	if maxSequence > 1<<sequenceBits-1 {
+		t.Errorf("unexpected max sequence: %d", maxSequence)
+	}
+}
+
+func Test_NewIDInParallel(t *testing.T) {
+	numCPU := runtime.NumCPU()
+	runtime.GOMAXPROCS(numCPU)
+
+	const numID = 10000
+	numGenerator := 1
+	if numCPU/2 > numGenerator {
+		numGenerator = numCPU / 2
+	}
+
+	generate := func(wg *sync.WaitGroup, partition []uint64) {
+		wg.Add(1)
```

Comment:
The test is failed to me with or without `-race` option. Removing this & adding `wg.Add(numGenerator)` after create the wait group fix it.

Reply:
From the document:
```
Note that calls with a positive delta that occur when the counter is zero must happen before a Wait.
Calls with a negative delta, or calls with a positive delta that start when the counter is greater than zero, may happen at any time.
Typically this means the calls to Add should execute before the statement creating the goroutine or other event to be waited for.
```
=> We need to call `wg.Add` before starting the go routines.

Reply:
Fixed